feat(cli): Setup command which uses an npm package

Josh-Walker-GM opened the pull request:
**Problem**
We want to allow users to run arbitary setup commands that come from npm packages. This is similar to the functionality of `npx` or `dlx` but we want to allow additional redwood specific behaviour. Specifically for now we want to ensure version compatibility between the setup package and the redwood version of the user's project.

**Notes**
I published the package `@joshgmw/rw-setup-example` as a means of testing the functionality in this PR.

**Changes**
1. Adds a `yarn rw setup package <npm-package>` command which will:
   1. Find the provided package on npm and check version compatibility
   2. Run `yarn dlx` with that package at the user project root

**Notes**
The UI of this isn't amazing but I don't really know what our consistent design for the CLI is. 

The naming of the command or where it sits in the command hierarchy is something I'm happy to change if we feel the further need to.

Josh-Walker-GM commented:
I marked this as `fixture-ok` because I think it is. I tried to run the rebuild fixture command locally but I don't think it's working all that well right now as it hit an error completely unrelated to any of my changes.

Tobbe commented:
> I tried to run the rebuild fixture command locally but I don't think it's working all that well right now as it hit an error completely unrelated to any of my changes.

I rebuilt the fixture yesterday. It failed on the first try. But running again it got all the way through

Josh-Walker-GM commented:
Okay @jtoar I think I've addressed the points you raised when we talked about this. It should be in a state for review now. 

There's less of the uncaught erroring out but I can't say the UI of this is great right now. I think this is likely just difficult because I feel we lack a consistent style or approach to CLI output at the moment.

When it comes to testing it out the package I published has a few different depencency ranges which you can see from the packument here: https://registry.npmjs.org/@joshgmw/rw-setup-example. That and manipulating your package.json @redwoodjs/core version should hopefully let you test the various cases.

Happy to iterate on this further if you feel things are quite right with it.

jtoar reviewed (commented):
@Josh-Walker-GM awesome work! thanks for providing a published package to test it out with. and love the tests! here's what i've got so far. will probably review one more time.

this kind of error seems unrecoverable right? if we continue, it's just going to fail. maybe we should just exit and skip the prompt? you can't run something that's not published:

<img width="832" alt="image" src="https://github.com/redwoodjs/redwood/assets/32992335/3fce5fdc-0bf6-4a35-97de-66283cd25a61">

jtoar commented on packages/cli/src/commands/setup/package/package.js:
```diff
@@ -0,0 +1,43 @@
+      type: 'boolean',
+      alias: 'f',
+    })
+    .option('options', {
+      default: [],
+      description: 'Options to forward on when running the package.',
+      type: 'array',
+    })
```
i think the array type here may not be what you expect looking at the docs here: https://github.com/yargs/yargs/blob/main/docs/tricks.md#arrays

Josh-Walker-GM replied:
Agreed that I shouldn't add new alternatives/conventions here that go against the existing ones - thanks for calling this out. 

Thoughts on using `--`? See: https://github.com/yargs/yargs/blob/main/docs/tricks.md#stop-parsing. I've pushed up a change to use this but if you'd prefer to switch back to `--fwd` I am happy to.

jtoar replied:
down for the double hyphen for now

Josh-Walker-GM commented on packages/cli/src/commands/setup/package/packageHandler.js:
```diff
@@ -0,0 +1,159 @@
+export async function handler({ npmPackage, force, options }) {
+  // Extract package name and version which the user provided
+  const isScoped = npmPackage.startsWith('@')
+  const packageName =
+    (isScoped ? '@' : '') + npmPackage.split('@')[isScoped ? 1 : 0]
+  const packageVersion = npmPackage.split('@')[isScoped ? 2 : 1] ?? 'latest'
```
Yeah it's not the most obvious code but I think it's correct. There was an immediately obvious solution to robustly splitting up the `@scope/package@version` so I quickly put that together. I think the tests covered it but happy to refactor to a better solution in the future if that's good?

Josh-Walker-GM commented:
@jtoar I've updated based on your feedback thanks! I was happy to swiftly exit when the package doesn't have the version or tag requested - as you said it'll error out a few seconds later anyway.
